### DISCUSSION TITLE
Remote test excecution with NUnit

### DISCUSSION THREAD

**Discussion by VentuzKarolStrozynski** [nunit/nunit#4250] [Q&A]
Hi there!

Is it possible to execute NUnit tests on a remote machine like described here under 'Test Engine Layer'?
https://docs.nunit.org/articles/nunit/technical-notes/nunit-internals/Architectural-Overview.html

I could not find anything about this feature.
To clarify: I want that e.g. Jenkins starts the NUnit tests on another machine like this is possible with mstest.exe , VS Test Controller and VS Test Agent.

Thanks in advance!

**Comment by stevenaw** [comment]
Hi @VentuzKarolStrozynski 
This _should_ indeed be possible provided all the networking and firewall rules in your setup permit it. Just to confirm: which test runner would you be using (nunit console, dotnet test, etc) and would that executable also be on the Jenkins machine or would it too be on a remote machine?

**Reply by VentuzKarolStrozynski** [reply]
Hi @stevenaw 
Thank you for your answer. 
I would use the NUnit Console and it could be on both Jenkins and the Remote Tester.
But in the NUnit Console command line options I do not see how to specify *where* the tests should be executed.
To give you a bigger picture: I'm looking for a replacement for MSTest.exe and the VS Test Controller/Agent collaboration which allowed to trigger a unit test on machine A but the execution would take place on machine B. It seems that this is not possible with .net 6 anymore respectively does MSTest V2 not support remote test execution. I need this remote execution because these tests are pretty GPU intensive and need huge resources located on the tester are required. Additionally I need to test on a machine without any Visual Studio installations.

**Comment by stevenaw** [comment]
Thanks for the explanation @VentuzKarolStrozynski 

The tests will run from whichever machine the console runner resides on, however I believe the console runner can itself remotely load test resources from another machine over, say, a UNC path to a network share.  Based on how you've described your setup, I think you'd likely want the console runner on the high-resource machine and then optionally the test dlls themselves on the same machine (though not strictly required, it could make any networking permissions easier). Then all you'd need is to ensure that your CI agent (Jenkins) can remotely execute the console runner.

This is all based on my understanding of the console runner which internally uses the nunit engine. There are others on the team more knowledgeable with the console runner than I. I'm hoping someone from @nunit/engine-team could help confirm this is all correct.

**Comment by CharliePoole** [comment]
This sounds like nunit/nunit-console#167, created five years ago and so far not carried out. It was once a part of the vision for NUnit 3.The console runner itself doesn't do a lot of heavy lifting and would therefore most likely run on the local machine, where the user would be able to interact with it. The issue itself has a very brief discussion of the possibilities for how the test assemblies would be loaded.

This is still a pretty desirable feature but will take a significant level of effort. I think it would be helpful for any future engine project lead to have @VentuzKarolStrozynski 's comments on that issue, rather than here in the framework project. But any work on it is going to be dependent on finding a new team leader and team to take over the engine after I leave at the end of the year.

The alternative, of course, is to continue to rely on the CI service to provision agents with sufficient resources to run the kinds of tests described in this issue.

**Reply by stevenaw** [reply]
Thanks @CharliePoole , I wasn't aware of that open issue on the console repo.
I've added the comments there
